rpi/src/drivers/fat32.rs

```diff
@@ -0,0 +1,239 @@
+
+        let mut boot_sector:Fat32BootSector = Default::default();
+        let buffer = unsafe{as_mut_buffer(&mut boot_sector)};
+        let _ = disk.read(bpb_lba_sector, buffer);
+
+        let fs_type_label = boot_sector.fs_type_label.clone();
+        if !(fs_type_label[0] == b'F' && fs_type_label[1] == b'A' && fs_type_label[2] == b'T'){
+            core::panic!("File system is not FAT");
+        }
+        if boot_sector.fat32_bpb.sectors_per_fat_16 != 0{
+            core::panic!("Detected FAT16 and not FAT32 file system");
+        }
+        if boot_sector.fat32_bpb.bytes_per_sector as u32 != disk.emmc.get_block_size(){
+            core::panic!("Currently dont support non 512 fat32 sectors");
+        }
+        let root_cluster = boot_sector.fat32_bpb.root_dir_first_cluster;
+        let fat_count = boot_sector.fat32_bpb.fats_count;
+        log::info!("root sector: {}", root_cluster);
+        log::info!("FAT count: {}", fat_count);
+
+        return Self { disk, boot_sector, lba_start_sector:bpb_lba_sector };
+    }
+
+    pub fn root_dir_list<const MAX:usize>(&mut self, offset:usize)->[FileEntry;MAX]{
+
+        let root_start_sector_index = self.get_cluster_start_sector_index(self.boot_sector.fat32_bpb.root_dir_first_cluster);
+        
+        let mut root_dir_files_count = 0;
+        let mut output_dir = [FileEntry::default();MAX];
+        let mut sector_offset = 0;
+        let mut discard = offset;
+
+        'search: loop{
+            let mut root_dir = [FatShortDirEntry::default();FAT_DIR_ENTRIES_PER_SECTOR];
+            let buffer = unsafe{as_mut_buffer(&mut root_dir)};
+            sector_offset += self.disk.read(root_start_sector_index + sector_offset, buffer);
+            for dir in root_dir{
+                if dir.file_name[0] == 0{
+                    break 'search;
+                }
+                else if dir.file_name[0] == 0xE5{
+                    continue;
+                }
+                if dir.attributes == ATTR_LONG_NAME{
+                    continue;
+                }
+                if discard > 0{
+                    discard -= 1;
+                    continue;
+                }
+
+                let mut filename:[u8;11] = [0;11];
+                filename[..8].copy_from_slice(&dir.file_name);
+                filename[8..11].copy_from_slice(&dir.file_extension);
+                let first_cluster_index = dir.first_cluster_index_low as u32 | ((dir.first_cluster_index_high as u32) << 16);
+                
+                output_dir[root_dir_files_count] = FileEntry{ name: filename, first_cluster_index, size: dir.file_size };
+                root_dir_files_count += 1;
+                if root_dir_files_count == MAX{
+                    break 'search;
+                }
+            }
+        }
+
+        return output_dir;
+    }
+
+    // Reads a file from the first FAT
+    pub fn read_file(&mut self, file_entry:&FileEntry, output:&mut [u32]){
```

Comment:
This function should receive a byte array and perhaps return a slice of the original buffer?